### ISSUE TITLE
Apply styling rules for commit messages

### ISSUE THREAD

**Issue by yukiisbored** [loklak/loklak_server#1076]
Currently the commit messages are pretty messy and doesn't have any consistency applied to them.

I think it's a good idea to have this implemented so every messages are clear.

**Comment by niccokunzmann** [comment]
I see the following roadmap for the issue:

1. [ ] Discuss what we need this for
2. [ ] Discuss the format
3. [ ] Document the format
4. [ ] Implement checkers

Contibution to (1)
---------------------

If we enforce a format, we better have a good reason for the burden of bureaucracy. I do not find a use in it but fear it will make it harder to enter the project and strengthen the maintainers' power.

Contribution to **(2)**
-------------------------

Proposal: A commit message SHOULD look like this:

1. Commits with a discussion or bigger changes need to reference the issue.
    ```
    <(a) description of the changes made> #<(b) issue number>
    <blank line with
    optional description after the blank line>
    ```
    Examples:
    - ```
        Using the note directive of sphinx #1024
        ```
    - ```
        apdapt path of docker file #1049
        
        The pull request #1050 is a precondition for this to make sense.
        ```
2. Commits which are assumed to be the final contribution to close the issue need to reference the issue like this:
    ```
    <(a) an optional description of the changes made> fixes #<(b) issue number>
    <blank line with
    optional description after the blank line>
    ```
    Examples:
    - `fixes #123`
    - `Move documentation site. fixes #123`
    - ```
        Move documentation site. fixes #123
        
        The documentation site is now moved to https://github.com/loklak/dev.loklak.org
        ```
3. Or, for small changes which are obvious, a pull-request should look like this:
    ```
    <(a) description of the changes made>
    <blank line>
    Problem: <Description of the problem>
    Solution: <Description of the solution>
    ```
    Examples:
    - ```
        spelling correted
        
        Problem: "desciption"
        Solution: "description"
        ```

Questions: 
- [ ] Should there be a merging policy?
  - [ ] Rebase all pull-requests to avoid merge commits?
- [ ] MUST this be enforced at the cost of loosing a contributor?
- [ ] Should we enforce maximum line length of 80 or 70 or something? (see linux)
- [ ] Should we use present tense or past tense?
- [ ] Is the contributing.md file to document this?

Requirements to close this issue after the discussion:
- [ ] This is documented
- [ ] The travis build checks the commit
- [ ] On Error: Travis outputs that the commit message needs to be changed and where to find
    - [ ] how to do that
    - [ ] what the format should be
- [ ] optional: add the travis description as a comment to the pull-request.

**Comment by yukiisbored** [comment]
@niccokunzmann here are my answers to your questions 
* The first line (aka the short log) should be shorter than 50, because that's the limit GitHub actually shows it.
  * Most likely, because merge commits are useless in my opinion
* Existing contributors must learn the new system. Just tell them to fix it and they will. 
* Probably to make command line viewing of `git log` much more pleasing to the eyes
* IMO present tense, because a commit is just a patch file that will do a change.
* Yes.

**Comment by yukiisbored** [comment]
I think we shouldn't enforce a formula/template because that's a bit hard to do especially for first timers.
I think it's better to make it a list of requirements instead.

**Comment by mariobehling** [comment]
We need to define a way as we proceed. Currently the project is relatively small and we can still define rules as we proceed. We are enforcing the best practices of FOSSASIA recently. This is already a step into the right direction. These "rules" are a start. Let's get more into detail if necessary at a later point and don't let us make things to elaborate for now. Closing this for now therefore.
